Unit testing tutorials and samples

jngrad opened the pull request:
Fixes #1377

Description of changes:
- **tests for tutorials and samples**
- **fixed broken samples and tutorial scripts**
- cleaned up tutorials and samples
- converted tutorial 8 to IPython


PR Checklist
------------
 - [x] Write tests (`/testsuite/scripts`)
 - [x] Document tests (`/doc/sphinx/testing.rst`)
 - [x] Integrate tests to CI (#1377):
   - update containers with missing Python packages:
     - nbconvert, matplotlib, scipy, mock (only Python2)
     - see error log at https://travis-ci.org/jngrad/espresso/builds/481848614
   - check containers have CMake version 3.7.0 or above (for test fixtures)
     - for example in Travis-CI job logs, system_info states CMake 3.9.2 when in reality it's 3.5.1
   - maybe schedule these tests differently (12 min total runtime)

pkreissl committed:
Cleanup visualization_constraints.py

pkreissl committed:
Directly set box_l

pkreissl committed:
Add # nopep8 to ignore where appropriate.

jngrad committed:
Remove Bjerrum length (#1664, #2428)

jngrad committed:
Fix tutorials and samples (porting #2433)

jngrad committed:
Clean up tutorial code

jngrad committed:
Clean up Markdown/LaTeX syntax in tutorials

jngrad committed:
Update text in tutorials

jngrad committed:
Handle vertical spacing in IPython notebooks

 - Consistent vertical spacing around LaTeX formula
 - Merge Markdown cells of paragraphs belonging to the same section
 - Replace 'new line' Markdown sequences (two spaces and a line feed,
   aka "  \n") by 'new paragraph' Markdown sequences (two line feeds,
   aka "\n\n"), except in bibliographies. The 'new line' sequence is
   not visible in IPython/Jupyter and won't create vertical spacing
   between the two text blocks it separates. It is only useful for
   condensing the space between bibliographic entries.

jngrad committed:
Add copyright header

jngrad committed:
Convert tutorial 8 to IPython notebook (#1572)

jngrad committed:
Update tutorial 8

 - Merged the 3 Python scripts into the notebook
   (they shared the same first 110 lines)
 - Adapted the interactive matplotlib code to run smoothly in
   IPython/Jupyter notebooks as well as in the classic Python
   terminal (need to toggle a global variable for the latter)

jngrad committed:
Restore tutorial 8 part on visualizer w/o plotting

jngrad committed:
Rename samples

jngrad committed:
Clean up samples (see #1845, #1661)

- Change instructions "from XYZ import *" to "import XYZ"
- Use argparse for complex command line interfaces
- Format messages with str.format()
- Use modern Python2.7 syntax
- Remove irrelevant/legacy code
- Wrap long strings into multi-line strings
- Move docstrings under the copyright header
- Give pickle files the .pkl extension

jngrad committed:
Prepare samples and tutorials for unit testing

- Store number of iterations in global variables
- Store numerical results to be tested in global variables
- CMake: copy tutorial scripts to the build directory

jngrad committed:
Fix NameError in try/except clause (Python 3)

In Python 2, "except Exception as e:" stores the exception in
a global variable "e". In Python 3, "e" is a local variable,
resulting in a "NameError: name 'e' is not defined" exception
if "e" is used later. This commit binds "e" to a class attribute.

jngrad committed:
Create dedicated Exception for missing features

jngrad committed:
Add tests for tutorials and samples (#1377)

jngrad committed:
Style patch

jngrad committed:
Document tests

pkreissl committed:
Merge branch 'samples_cleanup' into jngrad_fix-tutorials-samples-4.0-dev

jngrad committed:
Merge pull request #2 from pkreissl/jngrad_fix-tutorials-samples-4.0-dev

Merge pull request #2395 from pkreissl/samples_cleanup

Improve constrainst visualization sample; directly set box_l where appropriate

jngrad committed:
Fix issues in tests

jngrad committed:
Remove lj-demo test (too many dependencies)

jngrad committed:
Move IPython import to top-level project

jngrad committed:
Remove LaTeX dependency from tutorial tests

jngrad committed:
Style patch

jngrad committed:
Merge remote-tracking branch 'origin/fix-tutorials-samples-4.0-dev' into fix-tutorials-samples-4.0-dev

RudolfWeeber committed:
Merge branch 'python' of git://github.com/espressomd/espresso into pr2452

RudolfWeeber committed:
Merge regression

RudolfWeeber committed:
LB Syntax

RudolfWeeber committed:
lb syntax

RudolfWeeber commented:
@jngrad, it seems that the build system does not notice, when the original python files being checked have changed. The test still fails, even if the error in the underlying file has been fixed.

jngrad commented:
@RudolfWeeber I couldn't reproduce your observation. When I fix a sample, the build system immediately detects the change. It has to, because the first step of `check_samples` is to erase the local copy of the sample and start with a fresh one. This is done by the `local_samples` dependency, which is guaranteed to always be out-of-date thanks to `add_custom_target()`. It's wasteful to erase and copy the `samples` folder even if its content didn't change, but at least we avoid the situation you described. `check_samples` also depends on `pypresso`, so any change in the core will trigger the build system again. Didn't try on a tutorial, but the same CMake logic is used there as well. Could you please give me more details on what you did?

jngrad committed:
Fix LB and Constraints samples

jngrad committed:
Fix hanging tutorial tests

jngrad committed:
Fix incorrect error code handling

jngrad commented:
@RudolfWeeber the diamond polymer issue is fixed. I'm currently writing down the gitlab-ci.yaml section for the upcoming `cuda-tutorials` container. We probably don't want coverage enabled, right? The tutorials call many functions returning numerical results, but most of these results aren't checked. Also, these tutorials run on a different schedule.

jngrad committed:
Update CMake files for tutorial tests

* re-use CMake logic from /testsuite/python/CMakeLists.txt
* temporarily add cuda:tutorials container

jngrad committed:
Fix tag issue

jngrad commented:
The `cuda:tutorials` container works as expected. The next steps are:
1. reviewing this PR
2. edit: adding `schedules` to the `tutorials-samples` job in `.gitlab-ci.yml`
3. merging this PR
4. scheduling the tutorials tests

KaiSzuttor commented on .gitlab-ci.yml:
```diff
@@ -168,6 +168,17 @@ cuda-maxset:
     - linux
     - cuda
 
+cuda-tutorials:
```
why is the job named cuda-tutorials? we are not only checking cuda-tutorials but tutorials plus samples, right? Maybe just name it ```tutorials-samples```?

jngrad replied:
Good idea.

KaiSzuttor commented:
Why do you want to remove the container? You probably mean the job, but you can just add
``` yaml
  only:
    - schedules
```

to the job

jngrad commented:
Nevermind, didn't realize scheduled jobs were defined in the same file.

jngrad committed:
Change job title

KaiSzuttor commented on testsuite/scripts/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+configure_file(importlib_wrapper.py ${CMAKE_CURRENT_BINARY_DIR}/importlib_wrapper.py)
+configure_file(test_importlib_wrapper.py ${TEST_FILE_CONFIGURED_IMPORTLIB_WRAPPER})
+
+macro(PYTHON_SCRIPTS_TEST)
```
we should add a ```sample``` / ```tutorial``` prefix for these tests to distinguish them from the rest of the python tests

jngrad replied:
like so?
```
Test project /work/jgrad/espresso/build/testsuite/scripts/samples
      Start  1: importlib_wrapper
 1/59 Test  #1: importlib_wrapper ............................   Passed    0.60 sec
      Start  2: sample_save_checkpoint
 2/59 Test  #2: sample_save_checkpoint .......................   Passed    4.35 sec
      Start  3: sample_store_properties
 3/59 Test  #3: sample_store_properties ......................   Passed   10.78 sec
```
`importlib_wrapper` is neither a tutorial nor a sample, and won't appear in `check_python`

KaiSzuttor replied:
yes, looks good

jngrad committed:
Bugfix

jngrad committed:
Add prefix to sample and tutorial tests

KaiSzuttor commented:
@jngrad could you please the master and remove the WIP in the title if you think this is done

jngrad committed:
Merge branch 'python' into fix-tutorials-samples-4.0-dev

jngrad committed:
Add thermostat RNG seed

jngrad committed:
Fix regression introduced in f88df17c7c

jngrad commented:
@RudolfWeeber @KaiSzuttor ready for review!

jngrad committed:
Merge branch 'python' into fix-tutorials-samples-4.0-dev

jngrad committed:
Skip tests if no GPU

jngrad committed:
Test samples+tutorials in multiple configurations

jngrad committed:
Add espressomd.assert_features

jngrad commented:
Not sure what's up with the EK tutorial test. I had the same error on the CIP pool machine. On the surface it looked like a read/write race condition, where a file is written to disk and visible with `os.listdir('.')` but cannot be opened for reading just yet, however adding a delay between writing and reading with `time.sleep(0.5)` didn't help.

KaiSzuttor commented:
btw you shouldn't perform all the other checks in your job or better add your check to some existing CI job

jngrad committed:
Add logic for multiple imports

jngrad committed:
Change suffix

jngrad committed:
Workaround until CTest implements status "Skipped"

The CMake "PROPERTIES FIXTURES_REQUIRED" fixture only checks if the
dependency passed the test. When the dependency skipped a test, CTest
marks it as "Passed", because it does not implement statuses "Skipped"
nor "Partially Skipped". So a dependency that "Passed" the test may
not have created output files necessary for the dependent test, which
will fail instead of being skipped.

RudolfWeeber commented:
I think, the samples check with the empty myconfig failed, because the load_... sample are run before the store_... samples.

@KaiSzuttor, the tutorials are run in a special docker image at present. I think, we can run the one with the default myconfig on normal pull requests (but without check_python) and the ones with empty and maxset in the nightly schedule.;

jngrad commented:
@RudolfWeeber should be fixed now, see last commit message

jngrad committed:
Skip unit tests in samples+tutorials jobs

jngrad committed:
Fix regression introduced in 6f446d0

There was a name collision with an undeclared variable.

jngrad committed:
Make test less sensitive

KaiSzuttor commented:
Why should we check the tutorials and samples with different configs? We only want to make sure that interface changes don't break them. The CI already takes too long.

KaiSzuttor commented:
and why do we need an extra build for these checks? we can just exchange the image of an existing job and additionally check tutorials/samples

jngrad commented:
We use a dedicated docker image for the tutorials and samples, and we compile espresso with a different configuration in each tutorial job, so we cannot re-use artifacts.

jngrad committed:
Mitigate I/O race condition in tutorial 07 test

Unclear error on config "empty" causes the second import to fail upon
reading the processed python file that was just written to in the
importlib wrapper. Ordering the import order causes the test to be
skipped on the first import instead of the second import. This
doesn't change the test logic and seems to fix the issue.

jngrad committed:
Mitigate tutorial 07 test issue using importlib

jngrad commented:
The EK tutorial test should be fixed now.

fweik commented on doc/sphinx/testing.rst:
```diff
@@ -0,0 +1,419 @@
+
+.. code-block:: cmake
+
+   unit_test(NAME <mytest> SRC <mytest.c> [<../dependency1.c>, ...] [NUM_PROC <N>])
```
This is not up-to-date: the `unit_test` cmake function now supports `DEPENDS`, with which cmake targets can be added as dependencies for the test. The way you describe should not be used.

jngrad replied:
Text updated.

jngrad committed:
Update docs

RudolfWeeber commented:
@jngrad can you please mark the tutorials tests for default, empty and no-gpu as schedule only in gitlab-ci.yml.
Then, I think, this is ready.

jngrad committed:
Move the extra jobs to schedule

jngrad commented:
@RudolfWeeber you may mark the PR as reviewed

fweik commented:
bors r+
